src/tools/clippy/clippy_lints/src/significant_drop_in_scrutinee.rs

```diff
@@ -0,0 +1,235 @@
+use crate::FxHashSet;
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::get_attr;
+use clippy_utils::source::{indent_of, snippet};
+use rustc_errors::Applicability;
+use rustc_hir::intravisit::{walk_expr, Visitor};
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::{LateContext, LateLintPass, LintContext};
+use rustc_middle::ty::subst::GenericArgKind;
+use rustc_middle::ty::{Ty, TypeAndMut};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::Span;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Check for objects returned from function calls in a match scrutinee that have the attribute
+    /// `has_significant_drop`.
+    ///
+    /// ### Why is this bad?
+    /// The attribute `has_significant_drop` is added to types whose Drop impls have an important
+    /// side-effect, such as unlocking a mutex, making it important for users to be able to
+    /// accurately understand their lifetimes. When a temporary is returned in a function call in a
+    /// match scrutinee, its lifetime lasts until the end of the match block, which may be
+    /// surprising. It can also lead to deadlocks if the match scrutinee uses a function call that
+    /// returns a MutexGuard and then tries to lock again in one of the match arms, because the
+    /// MutexGuard will not be dropped until the end of the match block and thus will not unlock.
```

Comment:
```suggestion
    /// surprising.
    ///
    /// For `Mutex`es this can lead to a deadlock. This happens when the match scrutinee uses a function call that returns a
    /// `MutexGuard` and then tries to lock again in one of the match arms. In that case the
    /// `MutexGuard` in the scrutinee will not be dropped until the end of the match block and thus will not unlock.
```

Reply:
Thank you for the suggestion